src/services/penaltyGroups.js

```diff
@@ -74,43 +74,51 @@ export default class PenaltyGroup {
 		}
 	}
 
-	updatePenaltyGroupWithPayment(paymentInfo, callback) {
-		const getParams = {
-			TableName: this.penaltyGroupTableName,
-			Key: {
-				ID: paymentInfo.id,
-			},
-		};
-		const dbGet = this.db.get(getParams).promise();
+	async updatePenaltyGroupWithPayment(paymentInfo, callback) {
+		const { id, paymentStatus, penaltyType } = paymentInfo;
+		const oldPaymentStatus = paymentStatus === 'PAID' ? 'UNPAID' : 'PAID';
 
-		dbGet.then((data) => {
-			data.Item.PaymentStatus = paymentInfo.paymentStatus;
-			// TODO: Implement 'Enabled' flag (indicates whether or not a penalty group has been deleted)
-			data.Item.Hash = hashToken(paymentInfo.id, data.Item, true);
-			data.Item.Offset = getUnixTime();
+		try {
+			const penaltyGroup = await this._getPenaltyGroupById(id);
+			const penaltyDocs = await this._getPenaltiesWithIds(penaltyGroup.PenaltyDocumentIds);
+			const penaltiesByType = this._groupPenaltiesByType(penaltyDocs);
+			const penaltiesToUpdate = penaltiesByType[penaltyType];
+			const batchWriteParams = this._createUpdatePenaltiesPutParameters(
+				penaltiesToUpdate,
+				this.penaltyDocTableName,
+				paymentStatus,
+			);
+			// Update penalties
+			await this.db.batchWrite(batchWriteParams).promise();
+			// Check if all other penalties have been paid
+			const otherPenalties = penaltyDocs.filter(p => p.Value.penaltyType !== penaltyType);
+			const anyNotUpdated = otherPenalties.some(p => p.Value.paymentStatus === oldPaymentStatus);
+			if (!anyNotUpdated) {
+				// Update penalty group payment status if all penalties have been paid
+				penaltyGroup.PaymentStatus = paymentStatus;
+				// TODO: Implement 'Enabled' flag (indicates whether or not a penalty group was deleted)
+				penaltyGroup.Hash = hashToken(id, penaltyGroup, true);
+				penaltyGroup.Offset = getUnixTime();
+
+				const putParams = {
```

Comment:
As below, we're doing a put of the whole document, would it be easier to modify parts of it in place?